atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -905,7 +910,29 @@ public Map<Cell, byte[]> get(TableReference tableRef, Set<Cell> cells) {
                         tableRef,
                         cells,
                         uncached -> getInternal(
-                                "get", tableRef, uncached, immediateKeyValueService, immediateTransactionService));
+                                "get",
+                                tableRef,
+                                uncached,
+                                uncached.size(),
+                                immediateKeyValueService,
+                                immediateTransactionService));
+    }
+
+    @Override
+    public Map<Cell, byte[]> getWithExpectedNumberOfCells(
+            TableReference tableRef, Set<Cell> cells, int expectedNumberOfPresentCells) {
+        return getCache().get(tableRef, cells, uncached -> {
+            int cachedCells = cells.size() - uncached.size();
+            int numberOfCellsExpectingValuePostCache = expectedNumberOfPresentCells - cachedCells;
```

Comment:
Are we actually storing empty values in this cache (and on the one on `CachingTransaction`)?

If so, this might not be the correct thing to do, as we'd be wrongly lowering the number of expected present cells, leading to throwing expectations down the line or potentially returning empty values when we shouldn't have.

Reply:
Good spot! It's a bit messy:
- This cache uses lock watches, which does cache empty reads.
- The caching transaction (which I don't think is actually used in normal atlasdb usage) does also cache empty reads.

So yeah we might need a bit of refactoring here; as you note, the expression as written isn't quite right

Reply:
Refactored! Let me know how do you think things look like now.